Retail-Products samples

vasilypascal opened the pull request.

amanda-tarafa commented on retail/interactive-tutorial/RetailProducts.Samples.Tests/RetailProducts.Samples.Tests.csproj:
```diff
@@ -0,0 +1,28 @@
+
+  <ItemGroup>
+    <ProjectReference Include="..\RetailProducts.Samples\RetailProducts.Samples.csproj" />
+    <ProjectReference Include="..\Snippets\Snippets.csproj" />
```
Why is there this dependency here?

vasilypascal replied:
Because in Import Products samples we need to set up the data, and our setup/teardown test resources classes are situated inside "Snippets" project

vasilypascal replied:
But for now, I'll move Import samples, Import tests and Snippets to #1614 PR

amanda-tarafa commented on retail/interactive-tutorial/RetailProducts.Samples/UpdateProductSample.cs:
```diff
@@ -0,0 +1,146 @@
+    /// </summary>
+    /// <param name="length">The required length of alphanumeric string.</param>
+    /// <returns>Generated alphanumeric string.</returns>
+    private static string RandomAlphanumericString(int length)
```
I think you can substitute all of this for
```suggestion
Guid.NewGuid().ToString("N").Substring(0, 14).
```
Here and everywhere.

vasilypascal replied:
Agree, applied suggested improvement.

amanda-tarafa commented on retail/interactive-tutorial/RetailProducts.Samples/UpdateProductSample.cs:
```diff
@@ -0,0 +1,146 @@
+        var generatedProduct = new Product
+        {
+            Id = productId,
+            Name = $"projects/{projectId}/locations/global/catalogs/default_catalog/branches/default_branch/products/{productId}",
```
In general, we try that users don't need to know how resource names are formed.
```suggestion
            ProductName = new ProductName(projectId, "global", "default_catalog", "default_branch", productId),
```

vasilypascal replied:
Removed logic which generates another product object inside **Update** sample, so we no longer have this issue with the ProductName.

amanda-tarafa commented on retail/interactive-tutorial/RetailProducts.Samples/UpdateProductSample.cs:
```diff
@@ -0,0 +1,146 @@
+        Product updatedProduct = client.UpdateProduct(updateProductRequest);
+
+        Console.WriteLine("Updated product:");
+        Console.WriteLine(updatedProduct);
```
Consider being more explicit:
```suggestion
        Console.WriteLine($"Update title: {updatedProduct.Title}");
        ...
        ...
```

vasilypascal replied:
Sure. Improved the console outputs in all samples.

amanda-tarafa commented on retail/interactive-tutorial/RetailProducts.Samples.Tests/UpdateProductTest.cs:
```diff
@@ -0,0 +1,40 @@
+        [Fact]
+        public void TestUpdateProduct()
+        {
+            const string ExpectedProductTitle = "Updated Nest Mini";
+            const string ExpectedCurrencyCode = "EUR";
+            const float ExpectedProductPrice = 20.0f;
+            const float ExpectedProductOriginalPrice = 25.5f;
+            const Product.Types.Availability ExpectedProductAvailability = Product.Types.Availability.OutOfStock;
```
Since you are not reusing these, you can just use the literal values on the Asserts.

vasilypascal replied:
Agree. Resolved.

vasilypascal commented on retail/interactive-tutorial/Snippets/Snippets.csproj:
```diff
@@ -0,0 +1,17 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Moved to #1614 PR

vasilypascal commented on retail/interactive-tutorial/RetailProducts.Samples/UpdateProductSample.cs:
```diff
@@ -0,0 +1,146 @@
+        DeleteProductSample.DeleteRetailProduct(updatedProduct.Name);
+
+        return updatedProduct;
+    }
```
Agree. Moved all this logic inside **Tutorial** instead of **Sample** in every class.

vasilypascal commented on retail/interactive-tutorial/RetailProducts.Samples/UpdateProductSample.cs:
```diff
@@ -0,0 +1,146 @@
+        };
+
+        generatedProduct.Categories.Add("Updated Speakers and displays");
+        generatedProduct.Brands.Add("Updated Google");
```
Applied suggested style rule to all samples.

vasilypascal commented on retail/interactive-tutorial/RetailProducts.Samples/UpdateProductSample.cs:
```diff
@@ -0,0 +1,146 @@
+    /// <returns>Updated retail product.</returns>
+    public static Product UpdateRetailProduct(Product originalProduct, string projectId)
+    {
+        var productForUpdate = GenerateProductForUpdate(originalProduct.Id, projectId);
```
Totally agree with you. Modifying the existing **Product** object - looks like the right approach, so I've refactored the **Update** methods

vasilypascal commented on retail/interactive-tutorial/RetailProducts.Samples/UpdateProductSample.cs:
```diff
@@ -0,0 +1,146 @@
+        };
+
+        Console.WriteLine("Update product. request:");
+        Console.WriteLine(updateProductRequest);
```
Resolved.

vasilypascal commented:
@amanda-tarafa , Hello. I've refactored product samples by you suggestion and tested all samples and everything is working properly. Hope the whole picture looks better now. Waiting for your review. Thanks.
CC: @t-karasova

amanda-tarafa commented:
Thanks @vasilypascal I'll review as soon as possible, before the week is out.
I'm running tests here now.

amanda-tarafa commented:
@vasilypascal there's a failure that you might be able to see here: https://source.cloud.google.com/results/invocations/b37c726d-dc41-4460-a7e8-32e9e5c4a0c6/log, but just in case: 

```
X RetailProducts.Samples.Tests.RemoveFulfillmentInfoTest.TestRemoveFulfillmentPlaces [2m]

  Error Message:

   Assert.DoesNotContain() Failure

Found:    store0

In value: RepeatedField<String> ["store1", "store0"]

  Stack Trace:

     at RetailProducts.Samples.Tests.RemoveFulfillmentInfoTest.TestRemoveFulfillmentPlaces() in T:\src\github\dotnet-docs-samples\retail\interactive-tutorial\RetailProducts.Samples.Tests\RemoveFulfillmentInfoTest.cs:line 45
```

Note that these tests are run in parallel in Windows and Linux machines. It only failed on the Windows run so this is a flake. It might be one of two things:

- The change that the test makes takes a while to propagate (not sure if this may happen for the Retail API). If that's the case, you have to retry for a while (here and everywhere else where changes might take long to propagate). Let me know and I can show you how to retry with some helper clases we have on the repo.
- The test running on Ubuntu overrode the change made in Windows, in which case you need to create separate resources for each test run.

amanda-tarafa commented on retail/interactive-tutorial/RetailProducts.Samples/SetInventorySample.cs:
```diff
@@ -0,0 +1,126 @@
+        SetInventoryRequest setInventoryRequest = GetSetInventoryRequest(product);
+
+        ProductServiceClient client = ProductServiceClient.Create();
+        client.SetInventory(setInventoryRequest);
+
+        // This is a long running operation and its result is not immediately present with get operations,
+        // thus we simulate wait with sleep method.
+        Console.WriteLine("Set inventory. Wait 50 seconds:");
+        Console.WriteLine();
+
+        Thread.Sleep(50000);
```
The library does all of this for you, you just need to do as follows:

```suggestion
 // Make the request
            Operation<SetInventoryResponse, SetInventoryMetadata> response = client.SetInventory(setInventoryRequest);

            // Poll until the returned long-running operation is complete
            Operation<SetInventoryResponse, SetInventoryMetadata> completedResponse = response.PollUntilCompleted();
            // Retrieve the operation result
            SetInventoryResponse result = completedResponse.Result;
            
            Console.WriteLine(....);       
```
See the **Example** on this [section of the reference docs](https://cloud.google.com/dotnet/docs/reference/Google.Cloud.Retail.V2/latest/Google.Cloud.Retail.V2.ProductServiceClient#Google_Cloud_Retail_V2_ProductServiceClient_SetInventory_Google_Cloud_Retail_V2_Product_Google_Protobuf_WellKnownTypes_FieldMask_Google_Api_Gax_Grpc_CallSettings_) for more info.

vasilypascal replied:
Agree, this looks more generic, but after implementing this functionality, after every Api call "**response.PollUntilCompleted()**" I receive the next error message:

_Grpc.Core.RpcException: 'Status(StatusCode="PermissionDenied", Detail="The caller does not have permission", DebugException="Grpc.Core.Internal.CoreErrorDetailException: {"created":"@1647014879.849000000","description":"Error received from peer ipv4:142.250.186.170:443","file":"..\..\..\src\core\lib\surface\call.cc","file_line":1067,"grpc_message":"The caller does not have permission","grpc_status":7}")'_

However, the same method "**PollUntilCompleted()**" works fine for all Import samples. This error is received only for Add / Remove Fullfillment and SetInventory samples.

Maybe you have an idea of why is it happening to those samples?

vasilypascal replied:
FYI: all other comments are fixed, and looks like averything is working fine except this operation.

amanda-tarafa commented on retail/interactive-tutorial/RetailProducts.Samples/SetInventorySample.cs:
```diff
@@ -0,0 +1,126 @@
+    /// Perform inventory setting
+    /// </summary>
+    /// <param name="product">The actual product.</param>
+    public void PerformSetInventoryOperation(Product product)
```
Do you have these single line methods in the other samples we merged already? I don't remember these.

If the samples that we merged already have this, then it's OK to leave them in for consistency, but if they do not, then make these samples consistent with the ones we already merged.

vasilypascal replied:
No, we do not have these single line methods in "Search" samples, they just left after the latest refactoring, but I've removed them everywhere.

amanda-tarafa commented on retail/interactive-tutorial/RetailProducts.Samples/SetInventorySample.cs:
```diff
@@ -0,0 +1,126 @@
+    {
+        string[] placeIds = { "store1", "store2" };
+
+        product.FulfillmentInfo[0].Type = "pickup-in-store";
+        product.FulfillmentInfo[0].PlaceIds.Clear();
+        product.FulfillmentInfo[0].PlaceIds.AddRange(placeIds);
```
You can only access the element at 0 if the product had a fulfillment already.
Maybe consider showing how to deal with one that doesn't has a fullfillment?

```suggestion
        product.FulfillmentInfo.Clear();
        product.FulfillmentInfo.Add(new ....);
```

amanda-tarafa reviewed (changes requested):
I've left some comments that apply to all.

In particular for long running operations, you don't need to explicitly wait, the library does evrything for you, I've left an example there and a link to the library reference docs. I think this is why tests may be failing.

vasilypascal commented on retail/interactive-tutorial/RetailProducts.Samples/SetInventorySample.cs:
```diff
@@ -0,0 +1,126 @@
+    /// <param name="product">The actual product.</param>
+    private static void SetProductInventory(Product product)
+    {
+        SetInventoryRequest setInventoryRequest = GetSetInventoryRequest(product);
```
Resolved.

vasilypascal commented on retail/interactive-tutorial/RetailProducts.Samples/SetInventorySample.cs:
```diff
@@ -0,0 +1,126 @@
+        SetInventoryRequest setInventoryRequest = new SetInventoryRequest
+        {
+            Inventory = SetInventoryForProduct(product),
+            AllowMissing = true,
```
Resolved.

vasilypascal commented on retail/interactive-tutorial/RetailProducts.Samples/SetInventorySample.cs:
```diff
@@ -0,0 +1,126 @@
+            }
+        };
+
+        Console.WriteLine("Set inventory. request:");
```
Removed the dot in every sample.

vasilypascal commented on retail/interactive-tutorial/RetailProducts.Samples.Tests/UpdateProductTest.cs:
```diff
@@ -0,0 +1,51 @@
+                var sample = new UpdateProductSample();
+
+                // Update original product.
+                Product updatedProduct = sample.PerformUpdateProductOperation(originalProduct);
```
Adjusted.

vasilypascal commented on retail/interactive-tutorial/RetailProducts.Samples/UpdateProductSample.cs:
```diff
@@ -0,0 +1,117 @@
+    public static void PerformUpdateProductOperation()
+    {
+        var projectId = Environment.GetEnvironmentVariable("GOOGLE_PROJECT_ID");
+        var sample = new UpdateProductSample();
```
Removed.

vasilypascal commented on retail/interactive-tutorial/RetailProducts.Samples/UpdateProductSample.cs:
```diff
@@ -0,0 +1,117 @@
+        Product originalProduct = CreateProductSample.CreateRetailProduct(projectId);
+
+        // Update product.
+        Product updatedProduct = sample.PerformUpdateProductOperation(originalProduct);
```
Resovled.

vasilypascal commented on retail/interactive-tutorial/RetailProducts.Samples/UpdateProductSample.cs:
```diff
@@ -0,0 +1,117 @@
+    public Product PerformUpdateProductOperation(Product originalProduct)
+    {
+        return UpdateRetailProduct(originalProduct);
+    }
```
Removed the single line methods everywhere.

vasilypascal commented on retail/interactive-tutorial/RetailProducts.Samples/UpdateProductSample.cs:
```diff
@@ -0,0 +1,117 @@
+        Product updatedProduct = client.UpdateProduct(updateProductRequest);
+
+        Console.WriteLine($"Update title: {updatedProduct.Title}");
+        Console.WriteLine();
```
Fixed.

vasilypascal commented on retail/interactive-tutorial/RetailProducts.Samples/UpdateProductSample.cs:
```diff
@@ -0,0 +1,117 @@
+            AllowMissing = true
+        };
+
+        Console.WriteLine("Update product. request:");
```
Revoed the dot in every sample.

vasilypascal commented on retail/interactive-tutorial/RetailProducts.Samples/Tutorials/tutorial_set_inventory.md:
```diff
@@ -0,0 +1,148 @@
+﻿<walkthrough-metadata>
```
You are right. The were moved to another PR by @a-sahakyan

amanda-tarafa commented on retail/interactive-tutorial/RetailProducts.Samples/AddFulfillmentPlacesSample.cs:
```diff
@@ -0,0 +1,115 @@
+        {
+            Product = productName,
+            Type = "pickup-in-store",
+            AddTime = Timestamp.FromDateTime(RequestTimeStamp.AddMinutes(-1)),
```
Why are you substracting  1 minute? And actually, from the API docs it seems that it's not necessary to set the time in any of the requests? Are you sure you need to do this?

vasilypascal replied:
You are totally right, I have removed the request timestamp from all the requests from all the samples, except the one in **SetInventory**, as it's needed there for some testing purposes.

amanda-tarafa reviewed (changes requested):
Almost there:

- nits with some of the timestamps, how and where they are defined, but most importantly, if they are really needed
- questions about the huge Crud sample, is it really needed given that each of the CRUD operations is already being done by a smaller sample.
- question about the Create sample, as it is really 2 in one. If both are needed they need to be split in two files, else just keep the one that is needed.
- As for the permission error, I double checked and the account running these tests has all the required permissions. I guess you are getting the error locally as well, which would indicate a problem with the API itself. @chenLei1216 maybe we can chat about this internally? Also, I wonder if it has to do with the timestamp beeing specified and some clock sync issues, and we are getting Permissions denied instead of Invalid request? @vasilypascal can you try the failing samples without specifying a timestamp and see whether they continue failing?

vasilypascal commented on retail/interactive-tutorial/RetailProducts.Samples/SetInventorySample.cs:
```diff
@@ -0,0 +1,138 @@
+    private static SetInventoryRequest GetSetInventoryRequest(Product productWithInventory)
+    {
+        // The request timestamp.
+        DateTime requestTimeStamp = DateTime.Now.ToUniversalTime();
```
Resolved.

vasilypascal commented on retail/interactive-tutorial/RetailProducts.Samples/SetInventorySample.cs:
```diff
@@ -0,0 +1,138 @@
+        DateTime requestTimeStamp = DateTime.Now.ToUniversalTime();
+
+        // The out-of-order request timestamp:
+        // requestTimeStamp = DateTime.Now.ToUniversalTime().AddDays(-1);
```
This commented logic was removed. Also, SetInventory file was also removed, as per agreement with @tetiana-karasova

vasilypascal commented on retail/interactive-tutorial/RetailProducts.Samples/RemoveFulfillmentPlacesSample.cs:
```diff
@@ -0,0 +1,116 @@
+public class RemoveFulfillmentPlacesSample
+{
+    // The request timestamp
+    private static readonly DateTime RequestTimeStamp = DateTime.Now.ToUniversalTime();
```
Moved declaration of this variable to the method that builds the request, besides that, **RemoveFulfillmentPlaces** file was also removed, as per agreement with @tetiana-karasova

vasilypascal commented on retail/interactive-tutorial/RetailProducts.Samples/RemoveFulfillmentPlacesSample.cs:
```diff
@@ -0,0 +1,116 @@
+    private static readonly DateTime RequestTimeStamp = DateTime.Now.ToUniversalTime();
+
+    // The outdated request timestamp:
+    // RequestTimeStamp = DateTime.Now.ToUniversalTime().AddDays(-1);
```
This line was removed, and we no longer set timestamp, per agreement with @tetiana-karasova

vasilypascal commented on retail/interactive-tutorial/RetailProducts.Samples/CrudProductSample.cs:
```diff
@@ -0,0 +1,206 @@
+/// <summary>
+/// The crud product sample class.
+/// </summary>
+public class CrudProductSample
```
Initially, this sample was not in our plans, however, after the **python-samples** review, there was a suggestion to add this new sample, which does all CRUD operations just for visibility and as an example of how does all the flow look like, and to prove that all the flow works. 
@tetiana-karasova can you please confirm ?

amanda-tarafa replied:
OK, so I don't know about the suggestion in Python. In general, and I mean not just in .NET, but in all languages, we try to keep samples as simples as possible and showing only one operation, or at most operations that need to be performed one after the other.

If this was not in your plans to begin with, I would appreciate if we remove this one from .NET. It's basically duplicated code, harder to read for a user, and covered in full by individual samples. Thanks!

vasilypascal commented on retail/interactive-tutorial/RetailProducts.Samples.Tests/CreateProductTest.cs:
```diff
@@ -0,0 +1,42 @@
+            string projectId = Environment.GetEnvironmentVariable("GOOGLE_PROJECT_ID");
+
+            // Create product.
+            Product createdProduct = CreateProductSample.CreateRetailProduct(projectId);
```
Removed the unnecessary method, which created product without fulfillment.

vasilypascal commented on retail/interactive-tutorial/RetailProducts.Samples/AddFulfillmentPlacesSample.cs:
```diff
@@ -0,0 +1,115 @@
+    private static readonly DateTime RequestTimeStamp = DateTime.Now.ToUniversalTime();
+
+    // The outdated request timestamp:
+    // RequestTimeStamp = DateTime.Now.ToUniversalTime().AddDays(-1);
```
Removed timestamps.

vasilypascal commented:
@amanda-tarafa 
1) I have removed the request timestamp from all the requests from all the samples, except the one in **SetInventory**, as it's needed there for some testing purposes.
2) I've removed **AddFulfillment**, **RemoveFulfillment** and **SetInventory**, as per agreement with @tetiana-karasova, because the **PollUntilCompleted** operation is not working as expected, so there will be a separate PR for this samples.

amanda-tarafa commented on retail/interactive-tutorial/RetailProducts.Samples/DeleteProductSample.cs:
```diff
@@ -0,0 +1,80 @@
+        DeleteProductRequest deleteProductRequest = GetDeleteProductRequest(productName);
+
+        ProductServiceClient client = ProductServiceClient.Create();
+        client.DeleteProduct(deleteProductRequest);
+
+        Console.WriteLine($"Deleting product:");
+        Console.WriteLine($"Product {productName} was deleted");
```
```suggestion
        Console.WriteLine($"Deleting product:");
        client.DeleteProduct(deleteProductRequest);

        Console.WriteLine($"Product {productName} was deleted");
```

vasilypascal replied:
Fixed.

amanda-tarafa commented on retail/interactive-tutorial/RetailProducts.Samples/CreateProductSample.cs:
```diff
@@ -0,0 +1,129 @@
+    /// <returns>Create product request.</returns>
+    private static CreateProductRequest GetCreateProductRequest(Product productToCreate, string projectId)
+    {
+        string defaultBranchName = $"projects/{projectId}/locations/global/catalogs/default_catalog/branches/default_branch";
```
One other example of what I mentioned on the other PR about there being a type `BranchName` or similar that you can use to create these resource names.

vasilypascal replied:
I will refactor the way we are creating the Branches/Catalogs in a separate PR by your suggestion.

amanda-tarafa reviewed (changes requested):
Just a couple of nits, and a request to remove the Crud single sample. Otherwise this is ready.
Thanks!

vasilypascal commented:
@amanda-tarafa After discussion with @tetiana-karasova I've been told, that actually, the CRUD sample should still stay, because it was the requirement.
@tetiana-karasova can you please confirm or provide some additional context?

amanda-tarafa commented on retail/interactive-tutorial/RetailProducts.Samples/CrudProductSample.cs:
```diff
@@ -0,0 +1,206 @@
+// limitations under the License.
+
+// [START retail_crud_product]
+// Create product in a catalog using Retail API
```
```suggestion
```

amanda-tarafa commented on retail/interactive-tutorial/RetailProducts.Samples/CrudProductSample.cs:
```diff
@@ -0,0 +1,206 @@
+    {
+        return new Product
+        {
+            Id = Guid.NewGuid().ToString("N").Substring(0, 14),
```
Why do you have to add an ID here? Shouldn't you just use product name? Else, wouldn't you just create a new product (given that you have set allow missing to true)
```suggestion
```

amanda-tarafa reviewed (approved):
LGTM with just two minor comments.

I'll merge my suggestions myself and run tests, and if all is green we can merge after.

Do please create a new PR later for using the resource name helpers (branch name, catalog, etc.)

amanda-tarafa commented:
Error is unrelated, just PubSub timing out which can happen ocassionally.
Force merging...
